[Flake] MCP integration test: TestBookinfo_NamespaceCountsMatchResourceLists times out on list_or_get_resources

### Describe the bug

The backend MCP tools integration suite intermittently fails on TestBookinfo_NamespaceCountsMatchResourceLists with transient HTTP errors against Kiali’s MCP endpoint (POST /api/chat/mcp/list_or_get_resources).

This is not an assertion failure about bookinfo resource counts: the test dies on the first CallMCPTool call before comparing data.

Failure example
```
{"level":"warn","time":"2026-07-28T15:18:38Z","message":"MCP tool \"list_or_get_resources\" attempt 1/2 failed with transient error, retrying in 2s: POST http://172.18.255.70/kiali/api/chat/mcp/list_or_get_resources failed: Post \"http://172.18.255.70/kiali/api/chat/mcp/list_or_get_resources\": EOF"}
    bookinfo_resources_test.go:27:
        	Error: Received unexpected error:
        	            	POST .../list_or_get_resources failed: ... context deadline exceeded (Client.Timeout exceeded while awaiting headers)
--- FAIL: TestBookinfo_NamespaceCountsMatchResourceLists (140.72s)
FAIL	github.com/kiali/kiali/tests/integration/mcp_tools	158.001s
```
Observed pattern

- Attempt 1 fails with EOF (connection dropped).
- Attempt 2 fails with Client.Timeout exceeded while awaiting headers (90s timeout).
- Total test duration (~140s) matches a slow first failure + retry wait + full client timeout.

Failure repeats across runs → not a one-off CI glitch.

Likely cause

CI/network instability against the in-cluster Kiali MCP HTTP handlers (not a separate MCP process). A common aggravating factor: after EOF, the Go HTTP client can reuse a poisoned keep-alive connection, so the retry also fails/hangs.

Existing mitigations were limited:

MCPMaxAttempts = 2
fixed 2s retry wait
no CloseIdleConnections() before retry


### Expected Behavior

### What are the steps to reproduce this bug?

Example of the flake: https://github.com/kiali/kiali/actions/runs/30370840398/job/90317299515?pr=10106 

### Environment
*Learn about how to determine versions [here](https://kiali.io/docs/faq/general/#how-do-i-determine-what-version-i-am-running).*

- **Kiali version:**
- **Istio version:**
- **Kubernetes impl:** 
- **Kubernetes version:** 
- **Other notable environmental factors:**

--- a/tests/integration/mcp_tools/mcp_client.go
+++ b/tests/integration/mcp_tools/mcp_client.go
@@ -18,13 +18,14 @@ var mcpClient *MCPClient
 
 const (
 	MCPTimeout     = 90 * time.Second
-	MCPMaxAttempts = 2
+	MCPMaxAttempts = 4
 	MCPRetryWait   = 2 * time.Second
 )
 
 type MCPClient struct {
 	kialiURL   string
 	httpClient *http.Client
+	transport  *http.Transport
 }
 
 type MCPResponse struct {
@@ -42,14 +43,16 @@ func newMCPClient() *MCPClient {
 	if url == "" {
 		log.Fatalf("URL environment variable is required.")
 	}
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
 	return &MCPClient{
 		kialiURL: url,
 		httpClient: &http.Client{
-			Timeout: MCPTimeout,
-			Transport: &http.Transport{
-				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-			},
+			Timeout:   MCPTimeout,
+			Transport: transport,
 		},
+		transport: transport,
 	}
 }
 
@@ -69,6 +72,11 @@ func isTransientMCPError(err error) bool {
 		strings.Contains(msg, "EOF")
 }
 
+// mcpRetryWait returns exponential backoff: 2s, 4s, 8s for attempts 1..3.
+func mcpRetryWait(attempt int) time.Duration {
+	return MCPRetryWait * time.Duration(1<<(attempt-1))
+}
+
 func CallMCPTool(toolName string, args map[string]interface{}) (*MCPResponse, error) {
 	var lastErr error
 	for attempt := 1; attempt <= MCPMaxAttempts; attempt++ {
@@ -78,9 +86,12 @@ func CallMCPTool(toolName string, args map[string]interface{}) (*MCPResponse, er
 		}
 		lastErr = err
 		if attempt < MCPMaxAttempts && isTransientMCPError(err) {
+			// Drop poisoned keep-alive connections so the next attempt opens a fresh one.
+			mcpClient.transport.CloseIdleConnections()
+			wait := mcpRetryWait(attempt)
 			log.Warningf("MCP tool %q attempt %d/%d failed with transient error, retrying in %s: %v",
-				toolName, attempt, MCPMaxAttempts, MCPRetryWait, err)
-			time.Sleep(MCPRetryWait)
+				toolName, attempt, MCPMaxAttempts, wait, err)
+			time.Sleep(wait)
 			continue
 		}
 		return nil, err

